### PR TITLE
Time varying demography

### LINKED ISSUE #176: Time-varying demography
Time-varying demography, implemented via changin age-specific death rates over time.
Look into whether change demography need the reference to population size.
This is a seperate change to modelling variable population sizes over time.

### LINKED ISSUE #178: Demography vignette is broken
- plot not appearing

### PR DESCRIPTION
**Pull request by pwinskill**
Hi @giovannic,

I've had a first pass at implementing the time-varying demography as we discussed. In this PR I have:

- Created an age-group renderer to help visualise population demographies
- Adjusted the `set_demography()` function to allow a matrix of death rates and associated timesteps
- De-linked the dependency between demography and human population
- Updated the demography vignette to show age distributions and detail how a dynamic demography can be modelled.
- Added a check that when specifying custom demography one on the provided timesteps must be 0. This is to ensure we know what the baseline, starting demography is (this is unlike setting timesteps for interventions, where in the absence of a baseline coverage it is fine to assume it is 0).

These seem to be having the desired effect. Here is an example where I set a custom demog and crank up the deathrates in a few age groups in year 5.
image

This is a bit of a complex one for me and making the changes has raised a few questions it would be great to get your help with:

1. Do you agree that the n in a given age group output from the new `create_age_group_renderer()` should be named differntly and kept seperate from the similar variable created by the prevalence and incidecne renderers? I think it is best as it will allow seperate automation of post-processing for demographic specific outputs.
2. I wasn't sure how much of the population size functionality to leave in. I've left most of the functions for now incase we use them in the furture, but we could clean them up quite a bit?
3. The previous demog code used a utility function `match_timestep()` which wasn't working for me (it was indexing off by 1 essentially), I've added `match_last_timestep()` which seems to work. I wasn't sure if I was missing something or if that was a bug . Would be good if you could double check.
4. I've modified the relevant tests, but would really appreciate you checking them as they're somewhat complex for me to follow.
5. I'm getting an unexpected test fail locally (FOIM is consistent with equilibrium) - it doesn't seem like my changes should have impacted it, but not sure.

closes #178
closes #176

### CODE REVIEW

**Review by giovannic — CHANGES_REQUESTED** [review]
This is fantastic! I've left some comments on birthrate and tidyr stuff to clean up.

> I'm getting an unexpected test fail locally (FOIM is consistent with equilibrium) - it doesn't seem like my changes should have impacted it, but not sure.

Ohh I see this too, this seems out of scope for your PR. I've made an issue to look into it #182. You can skip the long running, analytical tests for now with `Sys.setenv(CI = 'true')`.

File: vignettes/Demography.Rmd [hunk 1]
```diff
@@ -28,17 +29,13 @@ year <- 365
...
   list(
     human_population = human_population,
-    model_seasonality = TRUE, # Let's try a bi-modal model
-    g0 = 0.28605,
-    g = c(0.20636, -0.0740318, -0.0009293),
-    h = c(0.173743, -0.0730962, -0.116019),
-    severe_incidence_rendering_min_ages = 2 * year,
-    severe_incidence_rendering_max_ages = 10 * year
+    age_group_rendering_min_ages = seq(0, 80, 5) * 365,
+    age_group_rendering_max_ages = seq(5, 85, 5) * 365
```
[the remark this answers is not in the document]
Comment (pwinskill) [reply vignettes/Demography.Rmd:38]: That's the increment in the sequence so yep

File: DESCRIPTION [hunk 2]
```diff
@@ -64,6 +64,7 @@ Suggests:
   DiagrammeR,
   cowplot,
   ggplot2,
+  tidyr,
```
[the remark this answers is not in the document]
Comment (pwinskill) [reply DESCRIPTION:67]: Have removed tidyr. Could work on removing ggplot from vignettes but it makes me a bit sad! Some of the others might be quite easy to get rid though

[the remark this answers is not in the document]
Comment (giovannic) [reply DESCRIPTION]: :( me too, ggplot is the only way I've learned to plot. I'm open to suggestions/experimentatation with build times

But just to be clear, that issue shouldn't affect this PR

**Review by giovannic — APPROVED** [review]
Fantastic! Thanks, sorry for the delay

**Pull request merged**